front/app/components/ConsentManager/Container.tsx

```diff
@@ -1,192 +1,129 @@
-`;
+// typings
+import { CategorizedDestinations, IPreferences } from './typings';
+import { TCategory } from './destinations';
 
 interface Props {
-  setPreferences: (arg: Record<string, unknown>) => void;
-  resetPreferences: () => void;
-  accept: () => void;
-  saveConsent: () => void;
-  isConsentRequired: boolean;
   preferences: IPreferences;
   categorizedDestinations: CategorizedDestinations;
+  isConsentRequired: boolean;
   onToggleModal: (opened: boolean) => void;
+  updatePreference: (category: TCategory, value: boolean) => void;
+  resetPreferences: () => void;
+  accept: () => void;
+  reject: () => void;
+  saveConsent: () => void;
 }
 
-interface State {
-  isDialogOpen: boolean;
-  isCancelling: boolean;
-}
-
-export default class Container extends PureComponent<Props, State> {
-  subscriptions: Subscription[] = [];
-
-  constructor(props: Props) {
-    super(props);
-    this.state = {
-      isDialogOpen: false,
-      isCancelling: false,
-    };
-  }
-
-  componentDidMount() {
-    this.subscriptions = [
-      eventEmitter
-        .observeEvent('openConsentManager')
-        .subscribe(this.openDialog),
-    ];
-  }
-
-  componentWillUnmount() {
-    this.subscriptions.forEach((subscription) => subscription.unsubscribe());
-  }
-
-  openDialog = () => {
-    this.props.onToggleModal(false);
-    this.setState({
-      isDialogOpen: true,
-    });
-  };
-
-  closeDialog = () => {
-    this.props.onToggleModal(true);
-    this.setState({
-      isDialogOpen: false,
-    });
-  };
-
-  handleCategoryChange = (category: string, value: boolean) => {
-    const { setPreferences } = this.props;
-
-    setPreferences({
-      [category]: value,
-    });
-  };
-
-  validate = () => {
-    let res = true;
-    const { preferences, categorizedDestinations } = this.props;
-    for (const category of Object.keys(categorizedDestinations)) {
-      if (categorizedDestinations[category].length > 0) {
-        res = res && !(preferences[category] === undefined);
-      }
-    }
-    return res;
-  };
-
-  handleSave = (e: FormEvent<any>) => {
-    e.preventDefault();
-
-    if (!this.validate()) {
-      return;
-    }
-
-    const { saveConsent } = this.props;
-
-    this.setState({
-      isDialogOpen: false,
-    });
-    saveConsent();
-  };
-
-  handleCancel = () => {
-    const { resetPreferences, isConsentRequired, preferences } = this.props;
-
+const Container = ({
+  preferences,
+  categorizedDestinations,
+  isConsentRequired,
+  onToggleModal,
+  updatePreference,
+  resetPreferences,
+  accept,
+  reject,
+  saveConsent,
+}: Props) => {
+  const [isDialogOpen, setIsDialogOpen] = useState(false);
+  const [isCancelling, setIsCancelling] = useState(false);
+
+  const openDialog = useCallback(() => {
+    onToggleModal(false);
+    setIsDialogOpen(true);
+  }, [onToggleModal]);
+
+  const closeDialog = useCallback(() => {
+    onToggleModal(true);
+    setIsDialogOpen(false);
+  }, [onToggleModal]);
```

Comment:
Yes- when refactoring from class-based to functional component, it's more accurate, as the class-based component methods are stable references, and this way these methods are also stable references.

Reply:
Interesting, the first time I hear about that. I don't know enough to make strong statements, nor is it important enough to have a discussion, but if I have to make a _guess_, I think they don't make much (if any) difference.

Reply:
I have experienced bugs with this before- the child component that the callback is passed to getting into infinite loops etc because the reference is not stable